### PR TITLE
[WIP] Make softmax accept a different scale/offset for input & output

### PR DESCRIPTION
**Pull request by tlepley-cadence**
This is a WIP PR related to issue #1512 

- Add `Type::isSimilar` to enable less strict type checking for quantized types
- Enables `Softmax `node and instruction to have a different type for the output with regard to the input

### CODE REVIEW

File: lib/Graph/Nodes.cpp [hunk 1]
```diff
@@ -228,7 +234,7 @@ static void verifyTanh(NodeValue src, NodeValue dest) {
 }
 
 static void verifySoftMax(NodeValue src, NodeValue dest) {
-  checkSameType(src, dest);
+  checkSimilarType(src, dest);
```
Comment (nadavrot) [inline lib/Graph/Nodes.cpp]: I think that isSimilar is not a very good name. Why not just compare the element type? It will work for integer and floats.

Comment (tlepley-cadence) [reply lib/Graph/Nodes.cpp]: I agree, the name is unclear :-) What I want to compare is not only the element type, but also the shape too. In other words, everything should be compared, except the quantization information (scale / offset)

Comment (jfix71) [reply lib/Graph/Nodes.cpp]: The shape is already checked in the next line (which could just be replaced with `checkSameShape()` to make that clear).

So you could just replace this `checkSimilarType()` with:
```
assert(src.getElementType() == dest.getElementType());
```

**Comment by rdzhabarov** [comment]
The direction looks good overall, I assume you are going to fix the quantization procedure for SoftMax next?

**Comment by rdzhabarov** [comment]
clicked the wrong button...

File: lib/Graph/Nodes.cpp (hunk 1 above)
Comment (tlepley-cadence) [reply lib/Graph/Nodes.cpp]: @nadavrot @jfix71 OK, I removed 'isSimilar', and like for some other nodes, I just check the equality for the shape like for arithmetic operation.

**Comment by tlepley-cadence** [comment]
@rdzhabarov I already changed `Quantization/Quantization.cpp` in this PR so that the quantization profile information is used for the output, what was not the case before.

File: include/glow/Graph/Graph.h [hunk 2]
```diff
@@ -270,7 +270,7 @@ class Function final : public Named {
   LogNode *createLog(llvm::StringRef name, NodeValue input);
 
   SoftMaxNode *createSoftMax(llvm::StringRef name, NodeValue input,
```
Comment (nadavrot) [inline include/glow/Graph/Graph.h]: @tlepley-cadence @rdzhabarov Most of our builders overload two versions of the builders: one with outTy and one without. Making outTy optional could be a good idea, but we need to be consistent. Not sure how to enforce the mandatory out type in a sane way.

File: tools/ClassGen/InstrGen.cpp [hunk 3]
```diff
@@ -147,8 +147,7 @@ int main(int argc, char **argv) {
   BB.newInstr("SoftMax")
       .addOperand("Dest", OperandKind::Out)
       .addOperand("Src", OperandKind::In)
-      .inplaceOperand({"Dest", "Src"})
-      .autoVerify(VerifyKind::SameType, {"Dest", "Src"})
+      .autoVerify(VerifyKind::NoVerify)
```
Comment (jfix71) [inline tools/ClassGen/InstrGen.cpp]: Can we do `VerifyKind::SameShape` here instead? I'd like to avoid `NoVerify` as much as possible.

File: include/glow/Graph/Graph.h (hunk 2 above)
Comment (rdzhabarov) [reply include/glow/Graph/Graph.h]: @nadavrot, I agree, let's do this consistently. I think there are two cases:
* Output shape calculations are based on the input shapes, and we do that inside the method which does not take outTy explicitly. The one which takes outTy just do the verification.
* We simply pass outTy as input.getType() to the other method

For the second case, we can eliminate one of the overloads and just check outTy for nullability. For the first case, we'd need to keep two overloads.

I think the way it's handled in the PR is the right way. Once this is merged I'll make sure we follow this practice elsewhere in the Graph.(h|cpp).

Comment (nadavrot) [reply include/glow/Graph/Graph.h]: Ok.

File: tools/ClassGen/InstrGen.cpp (hunk 3 above)
Comment (tlepley-cadence) [reply tools/ClassGen/InstrGen.cpp]: Done. I added the shape checking to `InstrGen.cpp`.

File: lib/Graph/Nodes.cpp [hunk 4]
```diff
@@ -228,8 +228,11 @@ static void verifyTanh(NodeValue src, NodeValue dest) {
 }
 
 static void verifySoftMax(NodeValue src, NodeValue dest) {
-  checkSameType(src, dest);
-  assert(src.dims() == dest.dims() && "Invalid shape");
+  if (src.getType()->isQuantizedType()) {
+    checkSameShape(src, dest);
```
Comment (jfix71) [inline lib/Graph/Nodes.cpp]: nit: Did you decide against also verifying src and dest have the same element type? Why not have it here?

Comment (tlepley-cadence) [reply lib/Graph/Nodes.cpp]: @jfix71 I actually did not add the element type equality check on purpose. For quantized networks, I think in general that we can't necessarily force the output element type to be the same as the input element type. I think this choice should be left to the quantization algorithm and it can depend on the target.

Comment (tlepley-cadence) [reply lib/Graph/Nodes.cpp]: Not testing the element type seems to be controversial and since I don't want to delay too much this PR that is primarily about solving a scale/offset issue, I reintroduced the element type checking for softmax. I'm currently doing some experiments about quantization and we can re-discuss more globally this element type problematic later in a different PR or issue.
As a side note, I've noticed that U8 is now handled in Glow not by introducing a new element type, but by acting on the S8 offset, so checking the element type equality between source and destination of nodes should not cause any issue in case the input and the output types of a node don't have the same signedness.

**Comment by rdzhabarov** [comment]
Thank you @tlepley! Assuming we can close https://github.com/pytorch/glow/issues/1512

Note, you can specify "closes #Issue" in the description of the PR and then once it's merged, the issue will be closed automatically.
